068-access-policies-in-jwt.md

```diff
@@ -0,0 +1,251 @@
+As the documentation states:
+
+> In addition to the Kafka custom resource configuration, the JAR file containing the custom authorizer class along with its dependencies must be available on the classpath of the Kafka broker.
+
+This proposal aims to add the JWT-based authorizer to Strimzi, so it can be selected in the configuration as
+another option apart from the simple and Keycloak ones, without having to implement a custom authorizer.
+
+## Motivation
+
+Having a generic solution is always preferable as it allows for less effort to be maintained.
+While we cannot achieve a perfect generic solution due to the nature of OAuth2, and how different
+identity providers are implemented, we can strive to create a solution as generic as possible.
+
+Passing additional data in the JWT seems to be supported by the majority of identity providers.
+Therefore, we can use this to pass the access policies in the JWT.
+This solution will work in most cases and with most identity providers, however it will not
+scale to infinity.
+It will, however, be decoupled from Keycloak and therefore be more generic.
+
+## Why JWT as an authorization mechanism?
+
+The official [JWT](https://jwt.io/introduction) lists some use cases:
+
+> Authorization: This is the most common scenario for using JWT. Once the user is logged in, each subsequent request will include the JWT, allowing the user to access routes, services, and resources that are permitted with that token. Single Sign On is a feature that widely uses JWT nowadays, because of its small overhead and its ability to be easily used across different domains.
+
+JWT supports custom claims, which can be used to pass additional claims/roles.
+For example [Spring Security](https://docs.spring.io/spring-authorization-server/reference/guides/how-to-custom-claims-authorities.html) can pass both claims and roles in the JWT,
+which can be used to authorize the user.
+
+All OAuth2 providers support custom claims in the JWT in one way or another.
+We have listed the most popular ones in the [Support for roles or claims in popular identity providers](#support-for-roles-or-claims-in-popular-identity-providers) section with links to the documentation
+on how custom claims or roles are supported.
+
+The concept of roles can be directly mapped to the concept of access policies in Kafka.
+
+## Proposal
+
+Identity providers support passing custom fields in the JWT token.
+Some providers allow the data type to be string, while others allow a list of strings.
+There are not many (if any) identity providers that allow a list of objects.
+
+Therefore, in this proposal we allow the ACLs to be passed as either.
+e.g. for a list of strings:
+```json
+{
+  "acls": [
+    "cluster_x:topic_1:read",
+    ":topic2:write"
+  ]
+}
+```
+
+and for a string:
+```json
+{
+  "acls": "cluster_x:topic_1:read,topic2:write"
+}
```

Comment:
why we need two different ways? Isn't it having `acls` as an array enough?

Reply:
I propose two ways in order to cover as many identity providers as possible. While passing `acls` as an array should be preferable, some identity providers support only passing strings, from what I could tell. Looking at [Google Firebase](https://firebase.google.com/docs/auth/admin/custom-claims), I cannot tell if they support passing arrays or just bools/strings/numbers. [OneLogin](https://developers.onelogin.com/api-docs/2/api-authorization/add-claim) is similar - they definitely support strings, but I cannot find information on supporting arrays.